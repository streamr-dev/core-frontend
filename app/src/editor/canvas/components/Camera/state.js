import isEditableElement from '$editor/shared/utils/isEditableElement'
import styles from './Camera.pcss'

function clamp(value, min, max) {
    return Math.max(min, Math.min(max, value))
}

function getParentMatching(el, selector) {
    if (!el) { return false }
    if (el.matches(selector)) { return el }
    return getParentMatching(el.parentElement, selector)
}

function scaleToFit({ bounds, fit, maxScale = 1, padding = 20 } = {}) {
    const totalPadding = 2 * padding
    const maxWidth = fit.width - totalPadding
    const maxHeight = fit.height - totalPadding
    return Math.min(maxScale, Math.min(maxWidth / bounds.width, maxHeight / bounds.height))
}

function centerBoundsTo({ bounds, fit, scale }) {
    // vertically & horizontally center content
    const offsetY = fit.height / 2
    const offsetX = fit.width / 2

    return {
        x: -((bounds.x + (bounds.width / 2)) * scale) + offsetX,
        y: -((bounds.y + (bounds.height / 2)) * scale) + offsetY,
    }
}

function toPrecision(v, precision = 0) {
    if (!precision) { return v }
    const p = 10 ** precision
    return Math.round((v * p)) / p
}

const defaultFit = {
    x: 0,
    y: 0,
    scale: 1,
}

function fitBoundsTo({ bounds, fit, padding }) {
    const scale = scaleToFit({
        bounds,
        fit,
        padding,
    })

    if (!scale) { return defaultFit }

    const { x, y } = centerBoundsTo({
        bounds,
        fit,
        scale,
    })

    return {
        x,
        y,
        scale,
    }
}

export function updateScaleState(s, { x, y, scale: newScale }) {
    const { scale: currentScale } = s
    if (!currentScale) { throw new Error(`invalid current scale: ${currentScale}`) }
    if (!newScale) { throw new Error(`invalid new scale: ${newScale}`) }

    // adjust for offset created by scale change
    const ratio = (1 - (newScale / currentScale))
    const x2 = s.x + ((x - s.x) * ratio)
    const y2 = s.y + ((y - s.y) * ratio)

    return {
        ...s,
        scale: newScale,
        x: x2,
        y: y2,
    }
}
const defaultCameraOptions = {
    scaleFactor: 0.1,
    scaleLevels: [
        0.1,
        0.25,
        0.5,
        1.0,
        1.25,
        1.50,
        2.0,
        3.0,
    ],
}

export function createCamera(opts = {}) {
    const { scaleFactor, scaleLevels } = Object.assign({}, defaultCameraOptions, opts)
    const minScale = scaleLevels[0]
    const maxScale = scaleLevels[scaleLevels.length - 1]
    return {
        scale: 1,
        x: 0,
        y: 0,
        elRef: opts.elRef || { current: undefined },
        scaleFactor,
        scaleLevels,
        minScale,
        maxScale,
    }
}

// clamps scale
export function setState(state, v) {
    const nextState = (typeof v === 'function') ? v(state) : v
    const { minScale, maxScale } = nextState
    return {
        ...nextState,
        scale: clamp(nextState.scale, minScale, maxScale),
    }
}

// updates scale of camera, centers scaling on x,y
export function updateScale(state, { x, y, delta }) {
    const { minScale, maxScale, scaleFactor } = state
    const factor = Math.abs(scaleFactor * (delta / 120))
    const actualScaleFactor = delta < 0 ? 1 + factor : 1 - factor
    return setState(state, (s) => (
        updateScaleState(s, {
            x,
            y,
            scale: clamp(s.scale * actualScaleFactor, minScale, maxScale),
        })
    ))
}

// changes current camera position
export function setPosition(state, { x, y }) {
    return setState(state, {
        ...state,
        x,
        y,
    })
}

function getBounds(state) {
    return state.elRef.current.getBoundingClientRect()
}

// set scale to value, centered on middle of camera el
export function setScale(state, value) {
    return setState(state, (s) => {
        const { width, height } = getBounds(s)
        return updateScaleState(s, {
            x: width / 2,
            y: height / 2,
            scale: toPrecision(value, 2),
        })
    })
}

// zoom in to next largest scale level
export function zoomIn(state) {
    const { scale, scaleLevels } = state
    const currentScale = toPrecision(scale, 2)
    const nextLevel = scaleLevels.find((v) => v > currentScale) || scale
    return setScale(state, nextLevel)
}

// zoom out to next largest scale level
export function zoomOut(state) {
    const { scale, scaleLevels } = state
    const currentScale = toPrecision(scale, 2)
    const nextLevel = scaleLevels.slice().reverse().find((v) => v < currentScale) || scale
    return setScale(state, nextLevel)
}

// changes current camera offset i.e. pans camera
export function pan(state, { x = 0, y = 0 } = {}) {
    return setState(state, (s) => ({
        ...s,
        x: s.x + x,
        y: s.y + y,
    }))
}

// fit camera to supplied bounds
export function fitBounds(state, opts) {
    return setState(state, (s) => ({
        ...s,
        ...fitBoundsTo(opts),
    }))
}

// fit supplied bounds to camera element bounds
export function fitView(state, opts) {
    const { width, height } = getBounds(state)
    return fitBounds(state, {
        ...opts,
        fit: {
            width,
            height,
        },
    })
}

// move bounds into center
export function centerBounds(state, { bounds }) {
    const { width, height } = getBounds(state)
    return setState(state, (s) => ({
        ...s,
        ...centerBoundsTo({
            bounds,
            fit: {
                width,
                height,
            },
            scale: s.scale,
        }),
    }))
}

export function areBoundsInView(state, { bounds, padding = 0 }) {
    const { x, y, scale } = state
    const { width, height } = getBounds(state)

    return (
        ((bounds.x * scale) + x) >= padding
        && ((bounds.y * scale) + y) >= padding
        && (((bounds.x + bounds.width) * scale) + x) < (width - padding)
        && (((bounds.y + bounds.height) * scale) + y) < (height - padding)
    )
}

// minimally move bounds into view, if not already in view
export function panIntoViewIfNeeded(state, { bounds, padding }) {
    const inView = areBoundsInView(state, {
        bounds,
        padding,
    })
    if (inView) { return }

    const { x, y, scale } = state
    const { width, height } = getBounds(state)

    const [overLeft, overTop, overRight, overBottom] = [
        padding - ((bounds.x * scale) + x),
        padding - ((bounds.y * scale) + y),
        (((bounds.x + bounds.width) * scale) + x) - (width - padding),
        (((bounds.y + bounds.height) * scale) + y) - (height - padding),
    ].map((v) => Math.max(0, v))

    // if too big to fit on screen rescale and center
    if ((overLeft > 0 && overRight > 0) || (overTop > 0 && overBottom > 0)) {
        return fitView(state, {
            bounds,
            padding,
        })
    }

    // prioritise top & left offsets over right & bottom
    const offsetX = overLeft > 0 ? overLeft : -overRight
    const offsetY = overTop > 0 ? overTop : -overBottom

    return setState(state, (s) => ({
        ...s,
        x: s.x + offsetX,
        y: s.y + offsetY,
    }))
}

// convert bounds in screen/camera coordinates to world coordinates
export function cameraToWorldBounds(state, bounds) {
    const { x, y, scale } = state
    const canvasX = (bounds.x - x) / scale
    const canvasY = (bounds.y - y) / scale
    return {
        x: canvasX,
        y: canvasY,
        width: bounds.width / scale,
        height: bounds.height / scale,
    }
}

// convert point in screen/camera coordinates to world coordinates
export function cameraToWorldPoint(state, point) {
    const { x, y } = cameraToWorldBounds(state, {
        ...point,
        width: 1,
        height: 1,
    })
    return {
        x,
        y,
    }
}

export function eventToWorldPoint(state, event) {
    const point = {
        x: 0,
        y: 0,
    }

    const { elRef } = state
    if (event.target === elRef.current) {
        // if target is camera el then we can just use event offsets
        point.x = event.offsetX
        point.y = event.offsetY
    } else {
        const { left, top } = getBounds(state)
        // otherwise need to calculate offset
        point.x = event.clientX - left
        point.y = event.clientY - top
    }

    const { x, y } = cameraToWorldBounds(state, {
        ...point,
        width: 1,
        height: 1,
    })
    return {
        x,
        y,
    }
}

export function shouldIgnoreEvent(state, event) {
    const { current: el } = state.elRef

    // do not ignore if has cameraControl class
    if (event.target.classList.contains(styles.cameraControl)) { return false }
    // always ignore if has noCameraControl class
    if (event.target.classList.contains(styles.noCameraControl)) { return true }
    // always ignore if target is editable
    if (isEditableElement(event.target)) { return true }

    // do not ignore if is within parent with cameraControl class
    const cameraControlEl = getParentMatching(event.target, `.${styles.cameraControl}`)
    // ignore if is within parent with noCameraControl class
    const noCameraControlEl = getParentMatching(event.target, `.${styles.noCameraControl}`)
    if (cameraControlEl && noCameraControlEl) {
        return cameraControlEl.contains(noCameraControlEl)
    }
    if (cameraControlEl) { return false }
    if (noCameraControlEl) { return false }

    return (
        // ignore if target not within camera el and isn't body
        !el.contains(event.target) && event.target !== document.body
    )
}
