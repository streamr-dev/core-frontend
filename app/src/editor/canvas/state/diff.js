import isEqual from 'lodash/isEqual'
import differenceWith from 'lodash/differenceWith'
import orderBy from 'lodash/orderBy'

const EMPTY = []

function getIsEqualIgnoring(ignoreKeys = new Set()) {
    return (objA, objB) => {
        if (objA === objB) { return true }
        const keysA = Object.keys(objA)
        /* eslint-disable no-continue */
        for (let i = 0; i < keysA.length; i += 1) {
            const key = keysA[i]
            if (ignoreKeys.has(key)) {
                // ignore module internals e.g. subcanvas module
                continue
            }
            if (objA[key] === objB[key]) {
                continue
            }
            const areEqual = isEqual(objA[key], objB[key])
            if (!areEqual) { return false }
        }
        /* eslint-enable no-continue */
        return true
    }
}

const MODULE_IGNORE_KEYS = new Set([
    'modules',
])

const CANVAS_IGNORE_KEYS = new Set([
    'created',
    'updated',
    'modules',
])

const isEqualModule = getIsEqualIgnoring(MODULE_IGNORE_KEYS)
const isEqualCanvasCheck = getIsEqualIgnoring(CANVAS_IGNORE_KEYS)

export function changedModules(canvasA, canvasB) {
    if (!canvasA || !canvasB) { return EMPTY }
    if (canvasA === canvasB || canvasA.modules === canvasB.modules) { return EMPTY }
    const modulesA = orderBy(canvasA.modules || [], 'hash')
    const modulesB = orderBy(canvasB.modules || [], 'hash')
    const AtoB = new Set(differenceWith(modulesA, modulesB, isEqualModule).map(({ hash }) => hash))
    const BtoA = new Set(differenceWith(
        // ignore items already flagged as changed
        modulesB.filter(({ hash }) => !AtoB.has(hash)),
        modulesA.filter(({ hash }) => !AtoB.has(hash)),
        isEqualModule,
    ).map(({ hash }) => hash))
    const result = Array.from(new Set([...AtoB, ...BtoA]))
    if (!result.length) { return EMPTY }
    return result
}

export function isEqualCanvas(canvasA, canvasB) {
    if (canvasA === canvasB) { return true }
    if (!canvasA || !canvasB) { return false }
    const hasChangedModules = changedModules(canvasA, canvasB)
    if (hasChangedModules.length) { return false }
    // don't re-check modules, ignore updated time
    return isEqualCanvasCheck(canvasA, canvasB)
}

