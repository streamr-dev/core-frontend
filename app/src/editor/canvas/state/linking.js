/**
 * Store links between objects (e.g. canvases) in local storage.
 *
 */

const LINK_PREFIX = 'Link'
function validateId(id) {
    if ((typeof id !== 'string' && typeof id !== 'number') || id === '') {
        throw new TypeError(`id must be non-empty string|number. Got ${typeof id}: ${id}`)
    }
}

export function getLink(parentId) {
    validateId(parentId)
    return window.localStorage.getItem(`${LINK_PREFIX}:${parentId}`)
}

export function link(parentId, childId) {
    validateId(parentId)
    validateId(childId)
    return window.localStorage.setItem(`${LINK_PREFIX}:${parentId}`, childId)
}

export function unlink(parentId) {
    validateId(parentId)
    window.localStorage.removeItem(`${LINK_PREFIX}:${parentId}`)
}
