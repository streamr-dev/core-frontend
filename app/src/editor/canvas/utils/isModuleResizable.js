// @flow

const RESIZABLE_MODULES = [
    'ChartModule',
    'CommentModule',
    'HeatmapModule',
    'MapModule',
    'LabelModule',
    'TableModule',
    'StreamrTextField',
    'ImageMapModule',
]

export default ({ jsModule, widget }: any) => (
    RESIZABLE_MODULES.includes(jsModule) || RESIZABLE_MODULES.includes(widget)
)
