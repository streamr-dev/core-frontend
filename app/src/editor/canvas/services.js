import axios from 'axios'

import Autosave from '$editor/shared/utils/autosave'
import { emptyCanvas } from './state'

const API = axios.create({
    headers: {
        'Content-Type': 'application/json',
    },
    withCredentials: true,
})

const getData = ({ data }) => data

const canvasesUrl = `${process.env.STREAMR_API_URL}/canvases`
const getModuleURL = `${process.env.STREAMR_URL}/module/jsonGetModule`
const getModuleTreeURL = `${process.env.STREAMR_URL}/module/jsonGetModuleTree`

const AUTOSAVE_DELAY = 3000

async function save(canvas) {
    return API.put(`${canvasesUrl}/${canvas.id}`, canvas).then(getData)
}

export const autosave = Autosave(save, AUTOSAVE_DELAY)

export async function saveNow(canvas, ...args) {
    if (autosave.pending) {
        return autosave.run(canvas, ...args) // do not wait for debounce
    }

    return save(canvas, ...args)
}

async function createCanvas(canvas) {
    return API.post(canvasesUrl, canvas).then(getData)
}

export async function create() {
    return createCanvas(emptyCanvas()) // create new empty
}

export async function moduleHelp({ id }) {
    return API.get(`${process.env.STREAMR_API_URL}/modules/${id}/help`).then(getData)
}

export async function duplicateCanvas(canvas) {
    const savedCanvas = await saveNow(canvas) // ensure canvas saved before duplicating
    return createCanvas(savedCanvas)
}

export async function deleteCanvas({ id }) {
    await autosave.cancel()
    return API.delete(`${canvasesUrl}/${id}`).then(getData)
}

export async function getModuleTree() {
    return API.get(getModuleTreeURL).then(getData)
}

export async function addModule({ id }) {
    const form = new FormData()
    form.append('id', id)
    return API.post(getModuleURL, form).then(getData)
}

export async function loadCanvas({ id }) {
    return API.get(`${canvasesUrl}/${id}`).then(getData)
}

async function startCanvas(canvas, { clearState }) {
    const savedCanvas = await saveNow(canvas)
    return API.post(`${canvasesUrl}/${savedCanvas.id}/start`, {
        clearState: !!clearState,
    }).then(getData)
}

async function startAdhocCanvas(canvas, options = {}) {
    const savedCanvas = await saveNow(canvas)
    const adhocCanvas = await createCanvas({
        ...savedCanvas,
        adhoc: true,
        settings: {
            ...savedCanvas.settings,
            parentCanvasId: canvas.id, // track parent canvas so can return after end
        },
    })
    return startCanvas(adhocCanvas, options)
}

export async function start(canvas, options = {}) {
    const { adhoc } = options
    if (!adhoc) { return startCanvas(canvas, options) }
    return startAdhocCanvas(canvas, options)
}

export async function stop(canvas) {
    return API.post(`${canvasesUrl}/${canvas.id}/stop`).then(getData)
}
