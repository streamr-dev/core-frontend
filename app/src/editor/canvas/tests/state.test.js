import { setupAuthorizationHeader, loadModuleDefinition } from '$editor/shared/tests/utils'

import * as State from '../state'
import * as Services from '../services'

import './utils'

const portMatcher = {
    id: expect.any(String),
    name: expect.any(String),
    longName: expect.any(String),
    type: expect.any(String),
}

describe('Canvas State', () => {
    let teardown

    beforeAll(async () => {
        teardown = await setupAuthorizationHeader()
    }, 60000)

    afterAll(async () => {
        await Services.deleteAllCanvases()
        await teardown()
    })

    let Clock

    beforeAll(async () => {
        Clock = await loadModuleDefinition('Clock')
        expect(Clock).toHaveProperty('id')
    })

    describe('emptyCanvas', () => {
        it('creates an empty canvas', () => {
            expect(State.emptyCanvas()).toMatchSnapshot({
                modules: [],
            })
        })
    })

    describe('updateCanvas', () => {
        it('does not add "undefined" the string as a key', () => {
            const canvas = State.updateCanvas(State.emptyCanvas())
            expect('undefined' in canvas).not.toBeTruthy()
        })
    })

    describe('addModule', () => {
        test('gives modules unique displayName', () => {
            let canvas = State.addModule(State.emptyCanvas(), Clock)
            canvas = State.addModule(canvas, Clock)
            canvas = State.addModule(canvas, Clock)
            const clock1 = canvas.modules[0]
            expect(canvas.modules[1]).toMatchObject({
                displayName: `${clock1.displayName} 02`,
            })
            expect(canvas.modules[2]).toMatchObject({
                displayName: `${clock1.displayName} 03`,
            })
        })
    })

    describe('getModule/findModule', () => {
        let canvas

        beforeAll(() => {
            canvas = State.addModule(State.emptyCanvas(), Clock)
        })

        test('findModule should find modules by matcher', () => {
            const clock = State.findModule(canvas, ({ name }) => name === 'Clock')
            expect(clock).toMatchObject({ name: 'Clock' })
        })

        test('getModule should find modules by hash', () => {
            const clock = State.findModule(canvas, ({ name }) => name === 'Clock')
            expect(State.getModule(canvas, clock.hash)).toBe(clock)
        })

        test('findModule should not error for missing modules', () => {
            expect(State.findModule(canvas, () => false)).toBeUndefined()
        })

        test('getModule should error for missing modules', () => {
            expect(() => State.getModule(canvas, 'missing')).toThrowError(State.MissingEntityError)
        })
    })

    describe('module port access', () => {
        let canvas
        let clock

        beforeAll(() => {
            canvas = State.addModule(State.emptyCanvas(), Clock)
            clock = State.findModule(canvas, ({ name }) => name === 'Clock')
        })

        describe('getModulePorts', () => {
            it('should get all module ports', () => {
                const modulePorts = State.getModulePorts(canvas, clock.hash)
                expect(modulePorts.length).toEqual(clock.inputs.length + clock.params.length + clock.outputs.length)
                modulePorts.forEach((port) => {
                    expect(port).toMatchObject(portMatcher)
                })
            })

            it('should error on missing module', () => {
                expect(() => State.getModulePorts(canvas, 'missing')).toThrowError(State.MissingEntityError)
            })
        })

        describe('getPort', () => {
            it('should get port', () => {
                const modulePorts = State.getModulePorts(canvas, clock.hash)
                modulePorts.forEach((port) => {
                    expect(State.getPort(canvas, port.id)).toBe(port)
                })
            })

            it('should error on missing port', () => {
                expect(() => State.getPort(canvas, 'missing')).toThrowError(State.MissingEntityError)
            })
        })

        describe('getAllPorts', () => {
            let canvasWithTwoModules
            let clocks

            beforeAll(async () => {
                canvasWithTwoModules = State.addModule(canvas, await loadModuleDefinition('Clock'))
                clocks = State.findModules(canvasWithTwoModules, ({ name }) => name === 'Clock')
            })

            it('should get all ports', () => {
                // get all expected ports via getModulePorts
                let expectedPorts = clocks.map((clock) => State.getModulePorts(canvasWithTwoModules, clock.hash))
                expectedPorts = new Set([].concat(...expectedPorts)) // flatten array into Set

                const allPorts = new Set(State.getAllPorts(canvasWithTwoModules))
                expect(expectedPorts).toEqual(allPorts)
            })
        })

        describe('ports', () => {
            let modulePorts

            beforeAll(() => {
                modulePorts = State.getModulePorts(canvas, clock.hash)
            })

            describe('hasPort', () => {
                it('should be true if has port', () => {
                    modulePorts.forEach((port) => {
                        expect(State.hasPort(canvas, port.id)).toBe(true)
                    })
                })

                it('should be false on missing port', () => {
                    expect(State.hasPort(canvas, 'missing')).toBe(false)
                })
            })

            describe('findModulePort', () => {
                it('should find ports by matcher', () => {
                    modulePorts.forEach((port) => {
                        expect(State.findModulePort(canvas, clock.hash, ({ id }) => id === port.id)).toBe(port)
                    })
                })

                it('should error on missing module', () => {
                    expect(() => State.findModulePort(canvas, 'missing', () => true)).toThrowError(State.MissingEntityError)
                })

                it('should not error on missing port', () => {
                    expect(State.findModulePort(canvas, clock.hash, () => false)).toBeUndefined()
                })
            })

            describe('getModuleForPort', () => {
                it('should get module for port', () => {
                    Object.values(modulePorts).forEach((port) => {
                        expect(State.getModuleForPort(canvas, port.id)).toBe(clock)
                    })
                })

                it('should error on missing port', () => {
                    expect(() => State.getModuleForPort(canvas, 'missing')).toThrowError(State.MissingEntityError)
                })
            })

            describe('arePortsOfSameModule', () => {
                let canvasWithTwoModules
                let clocks

                beforeAll(async () => {
                    canvasWithTwoModules = State.addModule(canvas, await loadModuleDefinition('Clock'))
                    clocks = State.findModules(canvasWithTwoModules, ({ name }) => name === 'Clock')
                })

                it('should be true for same module', () => {
                    const [port1, port2] = State.getModulePorts(canvasWithTwoModules, clocks[0].hash)
                    expect(State.arePortsOfSameModule(canvasWithTwoModules, port1.id, port2.id)).toBe(true)
                })

                it('should be false if not of same module', () => {
                    const [port1] = State.getModulePorts(canvasWithTwoModules, clocks[0].hash)
                    const [port2] = State.getModulePorts(canvasWithTwoModules, clocks[1].hash)
                    expect(State.arePortsOfSameModule(canvasWithTwoModules, port1.id, port2.id)).toBe(false)
                })

                it('should error on missing port', () => {
                    const [port] = State.getModulePorts(canvasWithTwoModules, clocks[0].hash)
                    expect(() => State.arePortsOfSameModule(canvasWithTwoModules, port.id, 'missing')).toThrowError(State.MissingEntityError)
                    expect(() => State.arePortsOfSameModule(canvasWithTwoModules, 'missing', port.id)).toThrowError(State.MissingEntityError)
                })
            })

            describe('setHistoricalRange', () => {
                const earlyDate = Date.now() - 9999999
                const lateDate = Date.now()
                it('should set endDate to beginDate when beginDate < endDate', () => {
                    let canvas = State.emptyCanvas()
                    const beginDate = earlyDate
                    const endDate = lateDate
                    canvas = State.setHistoricalRange(canvas, {
                        beginDate,
                    })
                    canvas = State.setHistoricalRange(canvas, {
                        endDate,
                    })
                    expect(canvas.settings.beginDate).toEqual(beginDate)
                    expect(canvas.settings.endDate).toEqual(endDate)
                })

                it('should set beginDate to endDate when beginDate < endDate & beginDate last set', () => {
                    const endDate = earlyDate
                    const beginDate = lateDate
                    let canvas = State.emptyCanvas()
                    canvas = State.setHistoricalRange(canvas, {
                        beginDate,
                    })
                    canvas = State.setHistoricalRange(canvas, {
                        endDate,
                    })
                    expect(canvas.settings.beginDate).toEqual(endDate)
                    expect(canvas.settings.endDate).toEqual(endDate)
                })

                it('should set endDate to beginDate when beginDate < endDate & endDate last set', () => {
                    const endDate = earlyDate
                    const beginDate = lateDate
                    let canvas = State.emptyCanvas()
                    canvas = State.setHistoricalRange(canvas, {
                        endDate,
                    })
                    canvas = State.setHistoricalRange(canvas, {
                        beginDate,
                    })
                    expect(canvas.settings.beginDate).toEqual(beginDate)
                    expect(canvas.settings.endDate).toEqual(beginDate)
                })

                it('should convert date strings to timestamps', () => {
                    let canvas = State.emptyCanvas()
                    const beginDate = earlyDate
                    const endDate = lateDate
                    canvas = State.setHistoricalRange(canvas, {
                        beginDate: new Date(beginDate).toISOString(),
                    })
                    canvas = State.setHistoricalRange(canvas, {
                        endDate: new Date(endDate).toISOString(),
                    })
                    expect(canvas.settings.beginDate).toEqual(beginDate)
                    expect(canvas.settings.endDate).toEqual(endDate)
                })

                it('should convert date strings to timestamps in updateCanvas, even without setHistoricalRange', () => {
                    let canvas = State.emptyCanvas()
                    const beginDate = earlyDate
                    const endDate = lateDate
                    canvas.settings.beginDate = new Date(earlyDate).toISOString()
                    canvas.settings.endDate = new Date(lateDate).toISOString()
                    canvas = State.updateCanvas(canvas)
                    expect(canvas.settings.beginDate).toEqual(beginDate)
                    expect(canvas.settings.endDate).toEqual(endDate)
                })
            })

            describe('updateModulePosition', () => {
                it('updates position', async () => {
                    let canvas = State.emptyCanvas()
                    canvas = State.addModule(canvas, await loadModuleDefinition('Equals'))
                    const [equals] = canvas.modules
                    const newPosition = {
                        top: Number.parseInt(equals.layout.position.top, 10) + 333,
                        left: Number.parseInt(equals.layout.position.left, 10) + 333,
                    }
                    canvas = State.updateModulePosition(canvas, equals.hash, newPosition)
                    const [equalsUpdated] = canvas.modules
                    expect(equalsUpdated.layout.position).toMatchObject({
                        top: `${newPosition.top}px`,
                        left: `${newPosition.left}px`,
                    })
                })
                it('throws on bad module hash', () => {
                    let canvas = State.emptyCanvas()
                    expect(() => {
                        canvas = State.updateModulePosition(canvas, 'not found', {
                            top: 0,
                            left: 0,
                        })
                    }).toThrow(State.MissingEntityError)
                })
            })

            describe('updateModuleSize', () => {
                it('updates size', async () => {
                    let canvas = State.emptyCanvas()
                    canvas = State.addModule(canvas, await loadModuleDefinition('Equals'))
                    const [equals] = canvas.modules
                    const newSize = {
                        width: Number.parseInt(equals.layout.width, 10) + 333,
                        height: Number.parseInt(equals.layout.height, 10) + 333,
                    }
                    canvas = State.updateModuleSize(canvas, equals.hash, newSize)
                    const [equalsUpdated] = canvas.modules
                    expect(equalsUpdated.layout).toMatchObject({
                        width: `${newSize.width}px`,
                        height: `${newSize.height}px`,
                    })
                })
                it('throws on bad module hash', () => {
                    let canvas = State.emptyCanvas()
                    expect(() => {
                        canvas = State.updateModuleSize(canvas, 'not found', {
                            width: 100,
                            height: 100,
                        })
                    }).toThrow(State.MissingEntityError)
                })
            })

            describe('{set,get}PortUserValue', () => {
                it('should coerce numberish values to numbers for Double type', async () => {
                    let canvas = State.emptyCanvas()
                    canvas = State.addModule(canvas, await loadModuleDefinition('Equals'))
                    const [equalsModule] = canvas.modules
                    // params[0] is 'tolerance', of Double type
                    // coerce empty string to default
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, ''))
                    const defaultValue = State.getPortDefaultValue(canvas, equalsModule.params[0].id)
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(defaultValue)
                    // handles commas
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, '2,3'))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(2.3)
                    // handles zero
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, '0'))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(0)
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, '0.0'))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(0)
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, 0))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(0)
                    // handles negative numbers
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, '-2.3'))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(-2.3)
                    // ignores whitespace
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, '2.3 '))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(2.3)
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, '2,3  '))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(2.3)
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, '  -2.3  '))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(-2.3)
                    // tries to parse a number
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, '2dasd'))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(2)
                    // Falls back to undefined if it cannot
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, equalsModule.params[0].id, 'dasd'))
                    expect(State.getPortUserValue(canvas, equalsModule.params[0].id)).toBe(defaultValue)

                    // test server accepts state
                    expect(State.updateCanvas(await Services.create(canvas))).toMatchCanvas(canvas)
                })
            })

            describe('getPortUserValueOrDefault', () => {
                it('gets value or default if value not set', async () => {
                    let canvas = State.emptyCanvas()
                    canvas = State.addModule(canvas, await loadModuleDefinition('MovingAverage'))
                    const [movingAverage] = canvas.modules
                    // gets value if set
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, movingAverage.params[0].id, 3))
                    expect(State.getPortUserValueOrDefault(canvas, movingAverage.params[0].id)).toBe(3)
                    // gets default if undefined
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, movingAverage.params[0].id, undefined))
                    expect(State.getPortUserValue(canvas, movingAverage.params[0].id)).toBe(0)
                    expect(State.getPortUserValueOrDefault(canvas, movingAverage.params[0].id)).toBe(0)
                    // gets default if empty string
                    canvas = State.updateCanvas(State.setPortUserValue(canvas, movingAverage.params[0].id, ''))
                    expect(State.getPortUserValue(canvas, movingAverage.params[0].id)).toBe(0)
                    expect(State.getPortUserValueOrDefault(canvas, movingAverage.params[0].id)).toBe(0)
                })
            })
        })
    })
})
