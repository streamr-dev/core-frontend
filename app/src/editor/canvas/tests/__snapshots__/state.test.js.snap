// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas State emptyCanvas creates an empty canvas 1`] = `
Object {
  "modules": Array [],
  "name": "Untitled Canvas",
  "settings": Object {
    "editorState": Object {
      "runTab": "#tab-realtime",
    },
  },
  "state": "STOPPED",
}
`;
